### PR TITLE
[Cherry-pick] Code for layerwise distillation

### PR DESCRIPTION
**Pull request by rahul-tuli**

### CODE REVIEW

File: src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
+            self._reset_cache()
```
Comment (anmarques) [inline src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py:362]: If I understand it correctly, _reset_cache is clearing the projection parameters. We want these parameters to be saved w/ the model and to be able to reload them

File: src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+@PyTorchModifierYAML()
+class PerLayerDistillationModifier(BaseDistillationModifier):
+    """
+    Adds a knowledge distillation loss based on the feature imitation loss.
+    A distillation_teacher module may be provided as a kwarg to
+    the Manager initialization and loss_update(loss) must be called before any
+    backwards pass in the integrated training flow.
+
+    If no teacher model is provided, then self-distillation will be used.
+    The feature difference between teacher and student can be weighted spatially
```
Comment (anmarques) [inline src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py:55]: The spatially weighing feature is not part of this PR

**Review by bfineran — APPROVED** [review]

**Review by anmarques — APPROVED** [review]
